hydro_lang/src/ir.rs

```diff
@@ -2277,6 +2301,113 @@ impl HydroNode {
+                    if let HydroNode::Persist { inner: input, .. } = input.as_mut() {
+                        (input, quote!('static))
+                    } else {
+                        (input, quote!('tick))
+                    };
+
+                let (input_ident, input_location_id) =
+                    input.emit_core(builders_or_callback, built_tees, next_stmt_id);
+
+                let input_type = *input.metadata().output_type.clone().unwrap().0;
+                let (key_type, value_type) = if let syn::Type::Tuple(tuple) = input_type {
+                    assert_eq!(
+                        tuple.elems.len(),
+                        2,
+                        "ReduceKeyedWatermark input must be a tuple of (key, value)."
+                    );
+                    (tuple.elems[0].clone(), tuple.elems[1].clone())
+                } else {
+                    panic!("ReduceKeyedWatermark input must be a tuple of (key, value).");
+                };
+
+                let (watermark_ident, watermark_location_id) =
+                    watermark.emit_core(builders_or_callback, built_tees, next_stmt_id);
+
+                let chain_ident = syn::Ident::new(
+                    &format!("reduce_keyed_watermark_chain_{}", *next_stmt_id),
+                    Span::call_site(),
+                );
+
+                let fold_ident =
+                    syn::Ident::new(&format!("stream_{}", *next_stmt_id), Span::call_site());
+
+                match builders_or_callback {
+                    BuildersOrCallback::Builders(graph_builders) => {
+                        assert_eq!(
+                            input_location_id, watermark_location_id,
+                            "ReduceKeyedWatermark inputs must be in the same location"
+                        );
+
+                        let builder = graph_builders.entry(input_location_id).or_default();
+                        // 1. Don't allow any values to be added to the map if the key <=the watermark
+                        // 2. If the entry didn't exist in the BTreeMap, add it. Otherwise, call f.
+                        //    If the watermark changed, delete all BTreeMap entries with a key < the watermark.
+                        // 3. Convert the BTreeMap back into a stream of (k, v)
+                        builder.add_dfir(
+                            parse_quote! {
+                                #chain_ident = chain();
+                                #input_ident
+                                    -> map(|x| (Some(x), None))
+                                    -> [0]#chain_ident;
+                                #watermark_ident
+                                    -> map(|watermark| (None, Some(watermark)))
+                                    -> [1]#chain_ident;
+
+                                #fold_ident = #chain_ident
+                                    -> fold::<#lifetime>(|| (::std::collections::BTreeMap::<#key_type, #value_type>::new(), None), {
```

Comment:
Why BTreeMap? Cause the removal is still O(nlog(n))?

Reply:
I was thinking so the flat_map order is more predictable

Reply:
But the output type is already `NoOrder`, so we don't care about predictability there...